File: android/core/player/src/main/java/com/mplayeraudio/core/player/PlaybackQueueBridge.kt
package com.mplayeraudio.core.player

import com.mplayeraudio.core.domain.musiclibrary.MusicProviderId
import com.mplayeraudio.core.domain.musiclibrary.TrackId
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.StateFlow

sealed interface PlayableSource {
    data class Remote(
        val provider: MusicProviderId,
    ) : PlayableSource

    data class Local(
        val uri: String,
    ) : PlayableSource
}

data class PlaybackQueueItem(
    val id: String,
    val trackId: TrackId,
    val source: PlayableSource,
    val title: String,
    val subtitle: String,
    val durationMs: Long,
    val artworkUri: String? = null,
)

enum class PlaybackPhase { Idle, Loading, Buffering, Playing, Paused, Failed, Ended }

sealed interface PlaybackError {
    val itemId: String?

    data class TrackUnavailable(override val itemId: String, val message: String) : PlaybackError
    data class StreamFailed(override val itemId: String, val message: String) : PlaybackError
    data class EngineCrashed(override val itemId: String?, val message: String) : PlaybackError
}

data class PlaybackQueueState(
    val queue: List<PlaybackQueueItem> = emptyList(),
    val currentIndex: Int? = null,
    val phase: PlaybackPhase = PlaybackPhase.Idle,
    val currentPositionMs: Long = 0L,
    val currentDurationMs: Long? = null,
    val bufferedPositionMs: Long = 0L,
    val playbackError: PlaybackError? = null,
) {
    val controlsEnabled: Boolean get() = queue.isNotEmpty()
    val isPlaying: Boolean get() = phase == PlaybackPhase.Playing
    val currentItem: PlaybackQueueItem? get() = currentIndex?.let(queue::getOrNull)
    val activeItemId: String? get() = currentItem?.id
    val playingItemId: String? get() = activeItemId.takeIf { isPlaying }
}

interface PlaybackQueueBridge : NowPlayingStripController {
    val playbackState: StateFlow<PlaybackQueueState>

    suspend fun replaceQueue(
        queue: List<PlaybackQueueItem>,
        startIndex: Int? = null,
        autoPlay: Boolean = false,
    )

    suspend fun playTrack(index: Int)

    fun acknowledgeError()

    fun shutdown()
}

interface PlayableUrlResolver {
    suspend fun getPlayableUrl(item: PlaybackQueueItem): String
}

File: ios/Core/Player/Sources/CorePlayer/PlaybackQueueBridge.swift
import CoreDomain
import Foundation

public enum PlayableSource: Equatable, Sendable {
    case remote(provider: MusicProviderId)
    case local(url: URL)

    public var descriptor: PlayableSourceDescriptor {
        switch self {
        case .remote(let provider):
            return .remote(provider: provider)

        case .local(let url):
            return .local(url: url)
        }
    }
}

public struct PlaybackQueueItem: Equatable, Sendable {
    public let id: String
    public let trackId: TrackId
    public let source: PlayableSource
    public let title: String
    public let subtitle: String
    /// Длительность трека в миллисекундах.
    public let durationMs: Int64
    public let artworkUri: String?

    public init(
        id: String,
        trackId: TrackId,
        source: PlayableSource,
        title: String,
        subtitle: String,
        durationMs: Int64,
        artworkUri: String? = nil
    ) {
        self.id = id
        self.trackId = trackId
        self.source = source
        self.title = title
        self.subtitle = subtitle
        self.durationMs = durationMs
        self.artworkUri = artworkUri
    }

    public var descriptor: PlaybackQueueItemDescriptor {
        PlaybackQueueItemDescriptor(
            id: id,
            trackId: trackId,
            source: source.descriptor
        )
    }
}

public struct PlaybackQueueState: Equatable, Sendable {
    public let queue: [PlaybackQueueItem]
    public let currentIndex: Int?
    public let isPlaying: Bool
    public let currentPositionMs: Int64
    public let controlsEnabled: Bool

    public init(
        queue: [PlaybackQueueItem] = [],
        currentIndex: Int? = nil,
        isPlaying: Bool = false,
        currentPositionMs: Int64 = 0,
        controlsEnabled: Bool? = nil
    ) {
        self.queue = queue
        self.currentIndex = currentIndex
        self.isPlaying = isPlaying
        self.currentPositionMs = currentPositionMs
        self.controlsEnabled = controlsEnabled ?? !queue.isEmpty
    }

    public var currentItem: PlaybackQueueItem? {
        guard let currentIndex else { return nil }
        return queue.indices.contains(currentIndex) ? queue[currentIndex] : nil
    }
}

public struct NowPlayingStripExternalState: Equatable, Sendable {
    public let title: String
    public let subtitle: String
    public let isPlaying: Bool
    public let currentPositionMs: Int64
    public let durationMs: Int64
    public let controlsEnabled: Bool

    public init(
        title: String = "",
        subtitle: String = "",
        isPlaying: Bool = false,
        currentPositionMs: Int64 = 0,
        durationMs: Int64 = 0,
        controlsEnabled: Bool = true
    ) {
        self.title = title
        self.subtitle = subtitle
        self.isPlaying = isPlaying
        self.currentPositionMs = currentPositionMs
        self.durationMs = durationMs
        self.controlsEnabled = controlsEnabled
    }
}

public protocol NowPlayingStripController: Sendable {
    func stripStateStream() -> AsyncStream<NowPlayingStripExternalState>
    func play() async
    func pause() async
    func skipNext() async
    func skipPrevious() async
    func seekTo(positionMs: Int64) async
}

public protocol PlaybackQueueBridge: NowPlayingStripController {
    func playbackStateStream() -> AsyncStream<PlaybackQueueState>
    func replaceQueue(queue: [PlaybackQueueItem], startIndex: Int?, autoPlay: Bool) async
    func playTrack(index: Int) async
}

public enum PlaybackDebugLog {
    public static let relativePath = "Library/Caches/player-debug.log"

    private static let queue = DispatchQueue(label: "com.mplayeraudio.playback-debug-log")
    private static let dateFormatter = ISO8601DateFormatter()
    private static let fileName = "player-debug.log"

    public static func reset() {
        queue.sync {
            try? FileManager.default.removeItem(at: fileURL())
        }
    }

    public static func record(
        category: String,
        message: String
    ) {
        let timestamp = dateFormatter.string(from: Date())
        let line = "\(timestamp) [\(category)] \(message)\n"

        queue.sync {
            let url = fileURL()
            let directoryURL = url.deletingLastPathComponent()
            try? FileManager.default.createDirectory(
                at: directoryURL,
                withIntermediateDirectories: true,
                attributes: nil
            )

            let data = Data(line.utf8)
            if FileManager.default.fileExists(atPath: url.path) {
                guard let handle = try? FileHandle(forWritingTo: url) else {
                    return
                }
                try? handle.seekToEnd()
                try? handle.write(contentsOf: data)
                try? handle.close()
                return
            }

            try? data.write(to: url, options: .atomic)
        }
    }

    private static func fileURL() -> URL {
        if let cachesURL = FileManager.default.urls(
            for: .cachesDirectory,
            in: .userDomainMask
        ).first {
            return cachesURL.appendingPathComponent(fileName)
        }

        return URL(fileURLWithPath: NSTemporaryDirectory()).appendingPathComponent(fileName)
    }
}
